### DISCUSSION TITLE
Type Narrowing Not Retained in Chained Array Methods with safeParse

### DISCUSSION THREAD

**Discussion by yuvals1** [colinhacks/zod#3145] [Q&A] [answered]
When using Zod's safeParse in conjunction with TypeScript's array filter method, the type narrowing does not persist to the subsequent map method.
The workaround involves using mySchema.parse(item) in map, which correctly infers the type, but it seems redundant to call parse after safeParse has already been used. 
Is there a more streamlined way to achieve this without the additional parsing line? Ideally, the type should be correctly narrowed without requiring a second parsing step. 

**Example:**
```typescript
// Incorrectly inferred type
const result = myArray
    .filter(item => mySchema.safeParse(item).success)
    .map(item => item.property);  // Type of 'item' is not correctly narrowed

// Correctly inferred type with type assertion
const correctResult = myArray
    .filter(item => mySchema.safeParse(item).success)
    .map(item => mySchema.parse(item))
    .map(item => item.myField); // Type of 'item' is correctly inferred 
```

**Comment by JacobWeisenburger** [comment, marked as answer]
The problem is not necessarily with Zod. It's with TypeScript. TypeScript is not able to infer the type of the array after the filter.

```ts
const data = [ 'foo', 123, 'bar', 42 ]
const result = data.filter( item => typeof item == 'string' )
// const result: (string | number)[]
```

However, you can use a type predicate to narrow the type of the array.

```ts
const data = [ 'foo', 123, 'bar', 42 ]
const result = data.filter( ( item ): item is string => typeof item == 'string' )
// const result: string[]
```

You can also use Zod to do the same thing.

```ts
const schema = z.string()
const data = [ 'foo', 123, 'bar', 42 ]
const result = data.filter( ( item ): item is string => schema.safeParse( item ).success )
// const result: string[]
```

More info:
https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
